### ISSUE TITLE
Clarify normalisation behaviour throughout

### ISSUE THREAD

**Issue by jarohen** [xtdb/crux#2917]
see https://github.com/xtdb/xtdb-team/blob/normalisation-readers-writers/normalisation-readers-writers.adoc#proposed-option

tl;dr, future state:

* RA and readers/writers are mostly oblivious to normalisation - they preserve exactly what the caller asks for
  * exception to the rule here being `getObject`, which should optionally take an `IKeyFn` to determine the struct keys returned - we considered this preferable to a second walk afterwards.
* So the callers of RA/readers/writers need to ensure that any (de-)normalisation is done beforehand/afterwards - i.e. the XTQL and SQL compilers.
  * Regardless of what case the user uses in their queries, the RA should be generated to yield normalised columns - it's the responsibility of the Arrow deserialisation code to denormalise the names as necessary.
* User supplies a `:key-fn` to `xt/q` - either `:kebab-case`, `:snake_case` or a function. If this is not provided, it defaults to `:kebab-case` in the Clojure API.
  * the user-visible change here is that we may appear not to honour the case that the user uses in their query, should they provide a different key-fn (or fall back to the default)
  * in other cases - for example, when the HTTP API is serving a request with `Accept: application/json` - the HTTP server may pass `:snake_case` to `xt/q` on the user's behalf.
